### PR TITLE
upgraded the build/CI to use net8.0

### PR DESCRIPTION
**Pull request by Yun-Ting**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/4437
Related to https://github.com/open-telemetry/opentelemetry-dotnet/pull/4868

## Changes

Upgraded the build/CI to use the .NET 8 SDK.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: build/Common.props [hunk 1]
```diff
@@ -20,6 +20,7 @@
     <TargetFrameworksForAspNetCoreInstrumentation>net7.0;net6.0;netstandard2.1;netstandard2.0</TargetFrameworksForAspNetCoreInstrumentation>
     <TargetFrameworksForGrpcNetClientInstrumentation>net6.0;netstandard2.1;netstandard2.0</TargetFrameworksForGrpcNetClientInstrumentation>
     <TargetFrameworksForPrometheusAspNetCore>net6.0</TargetFrameworksForPrometheusAspNetCore>
+    <TargetFrameworksForTests>net8.0;net7.0;net6.0</TargetFrameworksForTests>
```
Comment (xiang17) [inline build/Common.props:23]: Might fit better in Common.nonprod.props?

File: test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests/OpenTelemetry.Instrumentation.W3cTraceContext.Tests.csproj [hunk 2]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <Description>Unit test project for OpenTelemetry ASP.NET Core instrumentation for W3C Trace Context Trace</Description>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
-    <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == ''">net7.0;net6.0</TargetFrameworks>
+    <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == ''">$(TargetFrameworksForTests)</TargetFrameworks>
     <TargetFrameworks Condition="$(TARGET_FRAMEWORK) != ''">$(TARGET_FRAMEWORK)</TargetFrameworks>
```
Comment (reyang) [inline test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests/OpenTelemetry.Instrumentation.W3cTraceContext.Tests.csproj:7]:
```suggestion
    <TargetFrameworks $(TargetFrameworksForTests)</TargetFrameworks>
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.csproj [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
-    <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == ''">net7.0;net6.0</TargetFrameworks>
+    <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == ''">$(TargetFrameworksForTests)</TargetFrameworks>
     <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == '' AND $(OS) == 'Windows_NT'">$(TargetFrameworks);net462</TargetFrameworks>
     <TargetFrameworks Condition="$(TARGET_FRAMEWORK) != ''">$(TARGET_FRAMEWORK)</TargetFrameworks>
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.csproj:7]:
```suggestion
    <TargetFrameworks>$(TargetFrameworksForTests)</TargetFrameworks>
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.csproj [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
-    <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == ''">net7.0;net6.0</TargetFrameworks>
+    <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == ''">$(TargetFrameworksForTests)</TargetFrameworks>
     <TargetFrameworks Condition="$(TARGET_FRAMEWORK) == '' AND $(OS) == 'Windows_NT'">$(TargetFrameworks);net462</TargetFrameworks>
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.csproj:6]: extra this line to props file.

**Review by reyang — APPROVED** [review]
LGTM to unblock net8 support, I've added few comments but these can be addressed in a follow up PR.

File: test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests/OpenTelemetry.Instrumentation.W3cTraceContext.Tests.csproj (hunk 2 above)
Comment (reyang) [reply test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests/OpenTelemetry.Instrumentation.W3cTraceContext.Tests.csproj:7]: Check #4907.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.csproj (hunk 3 above)
Comment (reyang) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.csproj:7]: I sent a PR to fix these #4907.

File: build/Common.props (hunk 1 above)
Comment (reyang) [reply build/Common.props:23]: @xiang17 FYI https://github.com/open-telemetry/opentelemetry-dotnet/pull/4904#discussion_r1341771973

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 5]
```diff
@@ -912,6 +914,7 @@ void ConfigureTestServices(IServiceCollection services)
         Assert.Equal(4, numberofSubscribedEvents);
     }
 
+#if !NET8_0
```
Comment (CodeBlanch) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:917]: Any idea what's going on here? Was something fixed or broken in .NET 8?

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs [hunk 6]
```diff
@@ -14,7 +14,7 @@
 // limitations under the License.
 // </copyright>
 
-#if NET6_0_OR_GREATER
+#if NET6_0 || NET7_0
```
Comment (CodeBlanch) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs:17]: Same question as above. Any idea what happened here? Seems kind of odd something would change for .NET 8.

File: build/Common.props (hunk 1 above)
Comment (reyang) [reply build/Common.props:24]: This is now defined at Ln33 (introduced in #4907).

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs (hunk 6 above)
Comment (Yun-Ting) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs]: @CodeBlanch: 
I've updated the code to only not run when !OSX || !NET8_0.
https://github.com/open-telemetry/opentelemetry-dotnet/issues/4437#issuecomment-1741563896
I'm still trying to understand why.

Comment (reyang) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs]: What error did you see if we don't make the change? @Yun-Ting

Comment (Yun-Ting) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs]: 2023-09-29T22:56:50.2880408Z [xUnit.net 00:00:01.49]         /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs(1001,0): at OpenTelemetry.Instrumentation.AspNetCore.Tests.BasicTests.DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHandledInMiddleware()
2023-09-29T22:56:50.2888622Z [xUnit.net 00:00:01.49]         --- End of stack trace from previous location ---
2023-09-29T22:56:50.4395968Z warn: Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware[3]
2023-09-29T22:56:50.4396807Z       Failed to determine the https port for redirect.
2023-09-29T22:56:51.2553080Z   Failed OpenTelemetry.Instrumentation.AspNetCore.Tests.BasicTests.DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHandledInMiddleware [204 ms]
2023-09-29T22:56:51.2554395Z   Error Message:
2023-09-29T22:56:51.2555049Z    Assert.Equal() Failure: Values differ
2023-09-29T22:56:51.2555608Z Expected: 2
2023-09-29T22:56:51.2556137Z Actual:   1
2023-09-29T22:56:51.2556645Z   Stack Trace:
2023-09-29T22:56:51.2563218Z      at OpenTelemetry.Instrumentation.AspNetCore.Tests.BasicTests.DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHandledInMiddleware() in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:line 1001
2023-09-29T22:56:51.2564704Z --- End of stack trace from previous location ---

Comment (reyang) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs]: These seem to be flaky tests or transient error, which are not specific to TFMs.

File: test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs [hunk 7]
```diff
@@ -62,13 +62,13 @@ public void Tracer_StartRootSpan_BadArgs_NullSpanName()
             .Build();
 
         var span1 = this.tracer.StartRootSpan(null);
-        Assert.Null(span1.Activity.DisplayName);
+        Assert.True(string.IsNullOrEmpty(span1.Activity.DisplayName));
```
Comment (reyang) [inline test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs:65]: Why making this change?

Comment (CodeBlanch) [reply test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs:65]: Probably this: https://github.com/dotnet/runtime/pull/77746

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs (hunk 6 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs]: I merged this PR but please continue to look into this. I'm worried that we might have stumbled onto a .NET 8 bug and if so we should report it back to the team ASAP!

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj [hunk 8]
```diff
@@ -27,7 +27,7 @@
     <Compile Include="$(RepoRoot)\test\OpenTelemetry.Tests\Shared\TestEventListener.cs" Link="Includes\TestEventListener.cs" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net7.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:30]: We should use the latest version here https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Testing/8.0.0-rc.1.23421.29 for `.net8.0`

File: test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetry.Extensions.Hosting.Tests.csproj [hunk 9]
```diff
@@ -27,7 +27,7 @@
     </PackageReference>
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net7.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetry.Extensions.Hosting.Tests.csproj:30]: same as https://github.com/open-telemetry/opentelemetry-dotnet/pull/4876/files#r1343243487

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj (hunk 8 above)
Comment (reyang) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:30]: Curious, what does this mean?

Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:30]: Based on the compatible targetframework. https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Testing/8.0.0-rc.1.23421.29#supportedframeworks-body-tab.

Ideally, we should try get rid of this dependency from our test by following https://github.com/open-telemetry/opentelemetry-dotnet/blob/8ea6bb951caa40b11c36ac68317a5a62eb32192d/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs#L50

Comment (reyang) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:30]: I think we should take a different approach #4911.

Comment (reyang) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:30]: > Ideally, we should try get rid of this dependency from our test by following
> 
> https://github.com/open-telemetry/opentelemetry-dotnet/blob/8ea6bb951caa40b11c36ac68317a5a62eb32192d/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs#L50

Reducing dependency would also be great 👍

Comment (reyang) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:30]: Fixed by #4911.

File: test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetry.Extensions.Hosting.Tests.csproj (hunk 9 above)
Comment (reyang) [reply test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetry.Extensions.Hosting.Tests.csproj:30]: Fixed by #4911.

**Pull request merged**